src/runtime-rs/crates/shimv2/src/args.rs

```diff
@@ -0,0 +1,207 @@
+                &self.id, &self.namespace, &self.address, &self.publish_binary
+            ))));
+        }
+        if should_check_bundle && self.bundle.is_empty() {
+            return Err(anyhow!(Error::ArgumentIsEmpty("bundle".to_string())));
+        }
+
+        if Self::is_component_dangerous(&self.id) || Self::is_component_dangerous(&self.namespace) {
+            return Err(anyhow!(Error::DangenerousComponent(format!(
+                "id: {} namespace {}",
+                &self.id, &self.namespace
+            ))));
+        }
+
+        // Ensure `address` is a valid path.
+        let path = PathBuf::from(self.address.clone())
+            .canonicalize()
+            .context(Error::InvalidPath(self.address.clone()))?;
+        let md = path
+            .metadata()
+            .context(Error::FileGetMetadata(format!("{:?}", path)))?;
+        if !md.file_type().is_socket() {
+            return Err(Error::InvalidArgument).context("address is not socket");
+        }
+        self.address = path
+            .to_str()
+            .map(|v| v.to_owned())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Ensure `bundle` is a valid path.
+        if !self.bundle.is_empty() {
+            let path = PathBuf::from(self.bundle.clone())
+                .canonicalize()
+                .map_err(|_| Error::InvalidArgument)?;
+            let md = path
+                .metadata()
+                .map_err(|_| Error::InvalidArgument)
+                .context("get address metadata")?;
+            if !md.is_dir() {
+                return Err(Error::InvalidArgument).context("medata is dir");
+            }
+            self.bundle = path
+                .to_str()
+                .map(|v| v.to_owned())
+                .ok_or(Error::InvalidArgument)
+                .context("path to string")?;
+        }
+
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
+        if comp.is_empty() {
+            return true;
+        }
+        // only allow ascii alphanumeric character and '-', '_', '.' and '~'
+        !comp
+            .chars()
+            .all(|x| matches!(x, '0'..='9' | 'A'..='Z' | 'a'..='z' | '-' | '_' | '.' | '~'))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::os::unix::net::UnixListener;
+
+    #[test]
+    fn test_args_is_valid() {
+        let dir = tempfile::tempdir().unwrap();
+        let path = dir.path().to_path_buf();
+        let path = path.to_str().unwrap();
+        let bind_address = &format!("{}/socket1", path);
+        UnixListener::bind(bind_address).unwrap();
+
+        let mut arg = Args {
+            id: "1dfc0567".to_string(),
+            namespace: "ns1".to_owned(),
+            address: bind_address.to_owned(),
+            publish_binary: "containerd".to_string(),
+            socket: "socket".to_string(),
+            bundle: path.to_owned(),
+            debug: false,
+        };
+        arg.validate(false).unwrap();
```

Comment:
given how many assertions we have here, would it be possible add comments at different blocks of assertions so its clearer what we're testing for?

Reply:
Thanks for your review, has changed to table-driven-tests